private-identity-audit/wasm/src/lib.rs

```diff
@@ -0,0 +1,296 @@
+    pub fn initial_proofs_proofs(&self) -> Base64 {
+        self.proofs.clone()
+    }
+}
+
+#[wasm_bindgen]
+pub struct CommittedSetOutput {
+    verifier_secrets: Base64,
+    committed_uids: Base64,
+    challenge: Base64,
+}
+
+#[wasm_bindgen]
+impl CommittedSetOutput {
+    /// The committed set results.
+    #[wasm_bindgen(getter)]
+    pub fn verifier_secrets(&self) -> Base64 {
+        self.verifier_secrets.clone()
+    }
+
+    #[wasm_bindgen(getter)]
+    pub fn committed_uids(&self) -> Base64 {
+        self.committed_uids.clone()
+    }
+
+    #[wasm_bindgen(getter)]
+    pub fn challenge(&self) -> Base64 {
+        self.challenge.clone()
+    }
+}
+
+#[wasm_bindgen]
+pub struct ChallengeResponseOutput {
+    response: Base64,
+    committed_uids: Base64,
+}
+
+#[wasm_bindgen]
+impl ChallengeResponseOutput {
+    /// The initial proof results.
+    #[wasm_bindgen(getter)]
+    pub fn response(&self) -> Base64 {
+        self.response.clone()
+    }
+
+    #[wasm_bindgen(getter)]
+    pub fn committed_uids(&self) -> Base64 {
+        self.committed_uids.clone()
+    }
+}
+
+// ------------------------------------------------------------------------------------
+// -                                     Error Types                                  -
+// ------------------------------------------------------------------------------------
+#[wasm_bindgen]
+#[derive(Serialize)]
+pub enum WasmError {
+    GenerateInitialMessageError,
+    GenerateCommittedSetError,
+    GenerateChallengeResponseError,
+    ProofVerificationError,
+    DeserializationError,
+    Base64DecodingError,
+    DecryptionError,
+}
+
+impl From<WasmError> for JsValue {
+    fn from(e: WasmError) -> JsValue {
+        if let Ok(msg) = serde_json::to_string(&e) {
+            msg.into()
+        } else {
+            "Failed to serialized the error to string!".into()
+        }
+    }
+}
+
+type Fallible<T> = Result<T, JsValue>;
+
+// ------------------------------------------------------------------------------------
+// -                                     Public API                                   -
+// ------------------------------------------------------------------------------------
+
+/// Helper function to convert a Uid from Uuid to Scalar format.
+#[wasm_bindgen]
+pub fn convert_uuid_to_scalar(uuid: Base64) -> Fallible<Base64> {
+    match base64::decode(uuid) {
+        Ok(uuid) => {
+            let mut uuid_slice = [0u8; 16];
+            uuid_slice.copy_from_slice(uuid.as_slice());
+            Ok(base64::encode(
+                private_identity_audit::uuid_to_scalar(Uuid::from_bytes(uuid_slice)).as_bytes(),
+            ))
+        }
+        Err(_) => return Err(WasmError::DeserializationError.into()),
+    }
+}
+
+/// CDD provider's step #1:
+#[wasm_bindgen]
+pub fn generate_initial_proofs(cdd_claim: Base64) -> Fallible<InitialProofsOutput> {
+    let cdd_claim = decode::<CddClaimData>(cdd_claim)?;
+
+    let mut rng = OsRng;
```

Comment:
You raised a good comment last time about hiding the rng inside the function call. I know that I did it first, but is it a good idea to hide this rng inside the function call?
I hid it because I was not sure if the JS side can generate good random value. I also did not want to burden them with the task of generating good initial seeds.

What do you think?

Reply:
Yeah, as a general rule a crypto library should _never_ hide the fact that it is using an RNG underneath. It can particularly be a problem if you ship a code assuming there is an OSRng. I remember I debugged Schnorrkel with Mudit when he was running benchmarks, and it would just crash when it couldn't find the RNG at runtime.

For the FFI wrapper I replaced the RNG object with a seedable RNG and the API user provides the seed. Then I expect them to use the API with fresh seeds every time. We could do something similar here.

I saw that you got away with it in the MERCAT wrapper, so I didn't try harder :D

Reply:
But yeah, the real question is whether JS has access to good RNG or at least can read `dev/urandom`.

Reply:
@aafshar-poly I've convinced myself that using OSRng is good enough. See https://docs.rs/rand/0.5.5/rand/rngs/struct.OsRng.html. It uses each platform's random generation API (most of the time via `/dev/urandom`). The only catch is that, from what I've read, security of `Crypto.getRandomValues()` on browsers depends on whether the process's context is secure. But I don't understand JS enough to know whether we are running our processes in a secure context or not. See https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues :

>Don't use getRandomValues() to generate encryption keys. Instead, use the generateKey() method. There are a few reasons for this; for example, getRandomValues() is not guaranteed to be running in a secure context.

Reply:
Nice find. Based on my understanding, as long as the webapp uses https and uses recent version of TLS instead of deprecated version, it will considered a secure context. So from what you have quoted, if someone calls our WASM api from an app that is served over http, they'd be screwed since our OSRng is not guaranteed to be secure in that context.

So I guess it depends on what type of assurance/control we want to give to the users.
- If we use OSRng, they need to make sure to use https. If they use http, they will be insecure when using wasm without knowing it.
- If we instead receive a seed as input, we are putting the responsibility to them: they may pass a fixed seed all the time or they may use `generateKey` and we would not know...

Reply:
Yeah, I'm leaning towards keeping the OSRng in the Rust code, since it seems to be handling OS dependent RNG cases properly, and it might be too much to ask JS developers to do the same.

Reply:
Sounds good to me.